internal/docstore/mongodocstore/mongo.go

```diff
@@ -642,38 +460,74 @@ func (c *collection) unorderedBulkWrite(ctx context.Context, iactions []indexedA
 			err = gcerr.Newf(gcerr.Internal, nil, "bad action %+v", a)
 		}
 		if err != nil {
-			alerr = append(alerr, indexedError{ia.index, err})
+			errs[a.Index] = err
 		} else if m != nil { // m can be nil for a no-op update
 			models = append(models, m)
+			modelActions = append(modelActions, a)
+			newIDs = append(newIDs, newID)
 		}
 	}
+	if len(models) == 0 {
+		return nil
+	}
+	var reterrs []error
 	res, err := c.coll.BulkWrite(ctx, models, options.BulkWrite().SetOrdered(false))
 	if err != nil {
 		bwe, ok := err.(mongo.BulkWriteException)
 		if !ok { // assume everything failed with this error
-			return append(alerr, indexedError{-1, err})
+			return []error{err}
 		}
 		// The returned indexes of the WriteErrors are wrong. See https://jira.mongodb.org/browse/GODRIVER-1028.
 		// Until it's fixed, use negative values for the indexes in the errors we return.
 		for _, w := range bwe.WriteErrors {
-			alerr = append(alerr, indexedError{-1, gcerr.Newf(translateMongoCode(w.Code), nil, "%s", w.Message)})
+			reterrs = append(reterrs, gcerr.Newf(translateMongoCode(w.Code), nil, "%s", w.Message))
+		}
+		return reterrs
+	}
+	for i, newID := range newIDs {
+		if newID == nil {
+			continue
+		}
+		a := modelActions[i]
+		if err := a.Doc.SetField(c.idField, newID); err != nil {
+			errs[a.Index] = err
 		}
-		return alerr
 	}
 	if res.DeletedCount != nDeletes {
-		alerr = append(alerr, indexedError{-1,
-			gcerr.Newf(gcerr.NotFound, nil, "some delete failed (deleted %d out of %d)", res.DeletedCount, nDeletes)})
+		// Some Delete actions failed. It's not an error if a Delete failed because
+		// the document didn't exist, but it is an error if it failed because of a
+		// precondition mismatch. Find all the documents with revisions we tried to delete; if
+		// any are still present, that's an error.
+		c.determineDeleteErrors(ctx, models, modelActions, errs)
 	}
 	if res.MatchedCount+res.UpsertedCount != nNonCreateWrite {
-		alerr = append(alerr, indexedError{-1,
-			gcerr.Newf(gcerr.NotFound, nil, "some writes failed (replaced %d, upserted %d, out of total %d)", res.MatchedCount, res.UpsertedCount, nNonCreateWrite)})
-	}
-	for i, newID := range newIDs {
-		if err := iactions[i].action.Doc.SetField(c.idField, newID); err != nil {
-			alerr = append(alerr, indexedError{iactions[i].index, err})
+		reterrs = append(reterrs, gcerr.Newf(gcerr.NotFound, nil, "some writes failed (replaced %d, upserted %d, out of total %d)", res.MatchedCount, res.UpsertedCount, nNonCreateWrite))
+	}
+	return reterrs
+}
+
+func (c *collection) determineDeleteErrors(ctx context.Context, models []mongo.WriteModel, actions []*driver.Action, errs []error) {
```

Comment:
Just want to make sure mongo does report error on deleting items that don't exist.

Reply:
It doesn't, but docstore doesn't consider that an error.

Reply:
Would this be correctly reported once the issue you reported got fixed? So would the mismatched `nNonCreateWrite` number?